### PR TITLE
Unsubscribe a pulsar channel with pulsar ctl

### PR DESCRIPTION
**Pull request by xiaofan-luan**
Related #17530 
Use pulsar ctl to force unsubscribe a subscrition on datacoord or querycoord
Signed-off-by: xiaofan-luan <xiaofan.luan@zilliz.com>

### CODE REVIEW

**Comment by BUPTAnderson** [comment]
I have merge the patch, the datacoord crash again. @xiaofan-luan 
```
[2022/06/15 19:42:22.350 +08:00] [FATAL] [logutil.go:134] [panic] [recover="\"Failed to create consumer by-dev-rootcoord-dml_101, error = All attempts results:\
attempt #1:server error: ConsumerBusy: Exclusive consumer is already connected\
attempt #2:server error: ConsumerBusy: Exclusive consumer is already connected\
attempt #3:server error: ConsumerBusy: Exclusive consumer is already connected\
attempt #4:server error: ConsumerBusy: Exclusive consumer is already connected\
attempt #5:server error: ConsumerBusy: Exclusive consumer is already connected\
attempt #6:server error: ConsumerBusy: Exclusive consumer is already connected\
attempt #7:server error: ConsumerBusy: Exclusive consumer is already connected\
attempt #8:server error: ConsumerBusy: Exclusive consumer is already connected\
attempt #9:server error: ConsumerBusy: Exclusive consumer is already connected\
attempt #10:server error: ConsumerBusy: Exclusive consumer is already connected\
attempt #11:server error: ConsumerBusy: Exclusive consumer is already connected\
\""] [stack="github.com/milvus-io/milvus/internal/util/logutil.LogPanic
/opt/milvus/internal/util/logutil/logutil.go:134
runtime.gopanic
/usr/local/go/src/runtime/panic.go:965
github.com/milvus-io/milvus/internal/mq/msgstream.(*mqMsgStream).AsConsumerWithPosition
/opt/milvus/internal/mq/msgstream/mq_msgstream.go:174
github.com/milvus-io/milvus/internal/mq/msgstream.(*mqMsgStream).AsConsumer
/opt/milvus/internal/mq/msgstream/mq_msgstream.go:131
github.com/milvus-io/milvus/internal/mq/msgstream.(*PmsFactory).NewMsgStreamDisposer.func1
/opt/milvus/internal/mq/msgstream/mq_factory.go:97
github.com/milvus-io/milvus/internal/mq/msgstream.UnsubscribeChannels
/opt/milvus/internal/mq/msgstream/msgstream_util.go:30
github.com/milvus-io/milvus/internal/datacoord.(*ChannelManager).CleanupAndReassign
/opt/milvus/internal/datacoord/channel_manager.go:782
github.com/milvus-io/milvus/internal/datacoord.(*ChannelManager).processAck
/opt/milvus/internal/datacoord/channel_manager.go:621
github.com/milvus-io/milvus/internal/datacoord.(*ChannelManager).watchChannelStatesLoop
/opt/milvus/internal/datacoord/channel_manager.go:656"]
```

File: internal/mq/msgstream/mq_factory.go [hunk 1]
```diff
@@ -72,6 +75,34 @@ func (f *PmsFactory) NewQueryMsgStream(ctx context.Context) (MsgStream, error) {
 	return f.NewMsgStream(ctx)
 }
 
+func (f *PmsFactory) NewMsgStreamDisposer(ctx context.Context) func([]string, string) error {
+	return func(channels []string, subname string) error {
+		// try to delete the old subscription
+		cmdutils.PulsarCtlConfig.WebServiceURL = f.PulsarWebAddress
```
Comment (czs007) [inline internal/mq/msgstream/mq_factory.go:81]: pl prevent rw datarace.

File: internal/mq/msgstream/mq_factory.go [hunk 2]
```diff
@@ -72,6 +75,34 @@ func (f *PmsFactory) NewQueryMsgStream(ctx context.Context) (MsgStream, error) {
...
+			err = admin.Subscriptions().Delete(*topic, subname, true)
+			if err != nil {
+				log.Warn("failed to clean up subscriptions", zap.String("pulsar web", f.PulsarWebAddress),
+					zap.String("topic", channel), zap.Any("subname", subname), zap.Error(err))
+				// fallback to original way
+				msgstream, err := f.NewMsgStream(ctx)
+				if err != nil {
+					return err
+				}
+				msgstream.AsConsumer(channels, subname)
```
Comment (czs007) [inline internal/mq/msgstream/mq_factory.go:98]: Improve: AsConsumer(channels[i:], subname)

File: internal/mq/msgstream/mq_factory.go [hunk 3]
```diff
@@ -72,6 +75,33 @@ func (f *PmsFactory) NewQueryMsgStream(ctx context.Context) (MsgStream, error) {
...
 }
 
+func (f *PmsFactory) NewMsgStreamDisposer(ctx context.Context) func([]string, string) error {
+	return func(channels []string, subname string) error {
+		// try to delete the old subscription
+		admin := cmdutils.NewPulsarClient()
+		for _, channel := range channels {
+			topic, err := utils.GetTopicName(channel)
+			if err != nil {
+				log.Warn("failed to get topic name", zap.Error(err))
```
Comment (jaime0815) [inline internal/mq/msgstream/mq_factory.go:85]: Add channel name into log

File: internal/mq/msgstream/mq_factory.go [hunk 4]
```diff
@@ -72,6 +75,33 @@ func (f *PmsFactory) NewQueryMsgStream(ctx context.Context) (MsgStream, error) {
...
+				return retry.Unrecoverable(err)
+			}
+			err = admin.Subscriptions().Delete(*topic, subname, true)
+			if err != nil {
+				log.Warn("failed to clean up subscriptions", zap.String("pulsar web", f.PulsarWebAddress),
+					zap.String("topic", channel), zap.Any("subname", subname), zap.Error(err))
```
Comment (jaime0815) [inline internal/mq/msgstream/mq_factory.go:91]: the topic name is not channel

File: internal/mq/msgstream/msgstream.go [hunk 5]
```diff
@@ -77,4 +77,5 @@ type Factory interface {
 	NewMsgStream(ctx context.Context) (MsgStream, error)
 	NewTtMsgStream(ctx context.Context) (MsgStream, error)
 	NewQueryMsgStream(ctx context.Context) (MsgStream, error)
+	NewMsgStreamDisposer(ctx context.Context) func([]string, string) error
```
Comment (jaime0815) [inline internal/mq/msgstream/msgstream.go:80]: Why not add a Disposer to mqstream interface? the factory interface should use to msgstream of different type , unsubscribe should be a mq action

**Pull request merged**